### ISSUE TITLE
How to delay gatsby script execution in browser, and react hydration to after onLoad event?

### ISSUE THREAD

**Issue by teclone** [gatsbyjs/gatsby#25698]
Hello guys, this is a random question.

When I look at the build generated by gatsby, I see that all necessary scripts are preloaded in the head. Then, just before the body end tag, this scripts are appended with async directive. This is impressive, but....

Judging by the way async scripts behaves, it is likely that these scripts will start executing even before the browser had fired onLoad event, because they may have been preloaded at the time. If this eventually becomes the case, then the browser onLoad event will be delayed, till the scripts finish executing.

Then again, I do not know if gatsby checks for the onload event before booting up the scripts, client, react, webpack, etc.

What is the current implementation like? Why is it not possible to override/remove the generated scripts, or to even move the scripts to the head using any of the related ssr apis [onPreRenderHTML](https://www.gatsbyjs.org/docs/ssr-apis/#onPreRenderHTML) and [onRenderBody](https://www.gatsbyjs.org/docs/ssr-apis/#onRenderBody).

Someone can decide to disable javascript entirely on his website. just statically generated pages, lol. just saying. but this is a valid use case.

Well, the current arrangement of the scripts is very optimal for most use cases. You get preloading and parsing of the scripts, while the browser goes on to parse and render your html, but I am worried if gatsby does not check for onload event before booting things up because it is going to affect performance metrics very much for gatsby websites.

Conclusion.
If not currently done, all scripts in gatsby generated pages should only execute after the onLoad event has fired. A gatsby website is almost usable without scripting, so, it is best to run things, hydrate the client only after the onLoad event had fired.

Looking forward to some feedbacks and discussions :)

**Comment by herecydev** [comment]
> but I am worried if gatsby does not check for onload event before booting things up because it is going to affect performance metrics very much for gatsby websites

Why? Gatsby scores pretty much 100 out the box on all lighthouse scores.

> If not currently done, all scripts in gatsby generated pages should only execute after the onLoad event has fired.

How would that be done? As I understand you don't have any control over when the async script is run

**Comment by teclone** [comment]
hello @herecydev , gatsby does not score much based on the current web core vitals, lighthouse 6.0..

When you have an async script, the async script should go like this. 

```javascript
// all codes are here, etc
document.addEventListener('load', () => {
   startExecution();
});
```

This one is now bad

```javascript
/// all codes , functions, etc are here
startExecution();
```

The second one, will prevent dom on load event from firing, if the async script ends up getting downloaded and loaded before the end body tag.

I hope you understand now.

**Comment by teclone** [comment]
To prove that gatsby scripting is render-blocking based on the test on lighthouse 6.0, I have a gatsby site, deployed on s3. by loading the s3 URL directly, all scripts will not work, because of incorrect URLs. This makes it possible to see how the project scores without scripting.

Here is the link to the unscripted page, [directly from s3](https://s3.eu-central-1.amazonaws.com/dev.clacified.com.website/en/index.html)

Here is the link to the scripted page, [now on a proper domain](https://dev.clacified.com/en)

The one that is not scripted (like the scripts ends up not being loaded, 404 error) has this lighthouse score. Everything is practically green. [You can see the score here](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fs3.eu-central-1.amazonaws.com%2Fdev.clacified.com.website%2Fen%2Findex.html)
![Screenshot from 2020-07-12 11-24-05](https://user-images.githubusercontent.com/30493662/87243036-b7d7bf00-c432-11ea-8143-6c397f5682ea.png)

The second one, the same build, but on a proper domain, and scripting works has this score. [You can view the result here](https://developers.google.com/speed/pagespeed/insights/?url=https%3A%2F%2Fdev.clacified.com%2Fen)
![Screenshot from 2020-07-12 11-30-48](https://user-images.githubusercontent.com/30493662/87243085-359bca80-c433-11ea-9d68-02c76357060b.png)

This shows you how much the scripting impacts the page metrics. adding the scripts to the end body tag is not enough. because they have been preloaded on the html tag, and on fast internet, it would have been downloaded by the time the browser gets to it.

**Comment by teclone** [comment]
this shows the preloaded scripts in the head
![Screenshot from 2020-07-12 11-40-39](https://user-images.githubusercontent.com/30493662/87243248-d50d8d00-c434-11ea-8831-a6ea578df902.png)

and here is where they are lined just before the body end tag
![Screenshot from 2020-07-12 11-41-35](https://user-images.githubusercontent.com/30493662/87243256-e6569980-c434-11ea-96e0-74a1bcc69baf.png)

**Comment by herecydev** [comment]
Maybe I'm not following well but, we can't compare unscripted vs scripted that's not an option for most people. If you want no scripts that's possible with a plugin such as: https://www.gatsbyjs.org/packages/gatsby-plugin-no-javascript/ but it comes as a sacrifice to performance when browsing further pages.

So I've measured one of the gatsby starters (as that's independent), and the results are here: https://lighthouse-dot-webdotdevsite.appspot.com//lh/html?url=https%3A%2F%2Fgatsby-starter-blog-demo.netlify.app%2F

So that's got perfectly solid scores using lighthouse 6. I'm not sure how you can bring the TTI any lower with what you've suggested. Thoughts?

**Comment by teclone** [comment]
Hmm, judging by your result, it shows that the problem is from own scripts. I have checked my app-..... script bundle size, and it is enormously huge, something is wrong somewhere. I will investigate further and get back to you.

Thanks for this and for your time @herecydev

**Comment by teclone** [comment]
@herecydev , after some investigations, I found something weird, **Gatsby does not perform tree shaking for wrapPageElement and wrapRootElement components.**

Here is how I created out my test: I wanted to investigate why my script app bundle is very huge (about 890kb).

### Step 1

**I started from scratch and created just one single page, named test.tsx inside pages folder.
**here is the first content of the file. At this point, my app bundle size is 101.8kb (compressed but not zipped);**

```tsx
export default function () {
  return <div>Hello world</div>;
};
```

### Step 2
I added a WrapPage component inside `test.tsx`, something I would export inside `wrapPageElement.tsx` in later tests. After build, app bundle size stays the same (101.8kb)**

```tsx
import React from 'react';
import { WrapPage } from 'some-private-package';

const Render = () => {
  return <div>Hello world</div>;
};

export default function () {
  return (
    <div>
      <WrapPage
        element={<Render />}
        props={{ pageContext: { wrapPage: true } } as any}
        render={({ element }) => {
          return <div>{element}</div>;
        }}
      />
    </div>
  );
}
```

### Step 3:
I then moved the WrapPage component to a wrapPageElement.tsx file, which I then exported inside gatsby-browser.js and gatsby-browser.js files. Immediately, the app bundle size jumped to 312kb. from 101kb.

Here is the list of the file changes

`file wrapPageElement.tsx`
```tsx
import React, { FunctionComponent } from 'react';
import { WrapPage } from 'some-private-package';

export const wrapPageElement: FunctionComponent<any> = (props) => {
  return (
    <div>
      <WrapPage
        {...props}
        render={({ element, props }) => {
          return <div>{element}</div>;
        }}
      />
    </div>
  );
};
```

changed test.tsx file back to
```tsx
export default function () {
  return <div>Hello world</div>;
};
```

file gatsby-browser.js and file gatsby-ssr.js

```javacript
export {wrapPageElement} from './wrapPageElement.tsx';
```

**The interesting part of it, is that when I import WrapPage component directly from the file location inside the wrapPageElement like below**, it returns back to 101kb**
```
import {WrapPage} from 'private-packge/esm/components/WrapPage';
```

This also applies to WrapRootElement, where I import ApolloProvider and my ThemeProvider. Please any help is really appreciated.

**Comment by herecydev** [comment]
Are you certain it's not being split into a seperate webpack entrypoint?

**Comment by teclone** [comment]
Not at all, gatsby only generates one app js file

**Comment by teclone** [comment]
I could do a hangout call to show you what I am seeing here if you can please, this is weird

**Comment by herecydev** [comment]
> Not at all, gatsby only generates one app js file

Sure but it will also generate other js "bundles" if some criteria are met, so it might just be splitting the package into another bundle

**Comment by teclone** [comment]
But there is no need for it, as I am only importing a file from the package, not the whole package. I endup seeing unused codes inside the app bundle file, codes I never expected.

this is how it looks, the private package index.js file looks like this
```
export * from './helpers';
export * from './browser-helpers';
export * from './components/WrapPage';
```

the WrapPage component itself, imports from helpers like this
```
import { PageContext } from '../@types';
import React, { ReactNode } from 'react';
import { inPreview } from '../helpers';
```

the helpers index.js file inturn exports all of this
```
export * from './inPreview';
export * from './getPreviewPath';
export * from './compressGraphql';
export * from './parseQueryString';
and others.
```

Now, because the component WrapPage imports from `import { inPreview } from '../helpers';`, gatsby bundles everything into the app bundle. including getPreviewPath, compressGraphql, parseQueryString, and others.

If I change the import statement to `import { inPreview } from '../helpers/inPreview';` and also change `import {WrapPage} from 'private-package'` to `import {WrapPage} from 'private-package/esm/components/WrapPage'`, then it works as expected

**Comment by herecydev** [comment]
Not sure then, probably best to open another issue with a reproducable example

**Comment by teclone** [comment]
ok, @herecydev , I will do that

**Comment by teclone** [comment]
@herecydev , we can close this issue, I found the solution. To support complete tree shaking in npm libraries, you need to inform webpack that your codes has no side-effects, and that any unused code should be pruned using the sideEffects directive.

[More information is here for documentation purposes](https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free).

Thank you so much for your time. I have added this to the private package, and it now works as I expect
